File: update.py
#!/usr/bin/env python3
"""
Check for new Cursor releases and update PKGBUILD accordingly.

Cursor doesn't have a proper release API, but their download URLs follow a
predictable pattern. We probe their update endpoint to discover the latest
version and commit hash, then patch the PKGBUILD.

Exit codes:
  0 = updated (or already up-to-date with --check)
  1 = error
  2 = already up-to-date (with --update)
"""

import argparse
import hashlib
import json
import re
import subprocess
import sys
import tempfile
import urllib.request
from pathlib import Path


# Cursor's update check endpoint (same one the app uses)
UPDATE_URL = "https://api2.cursor.sh/updates/api/update/linux-x64/stable/latest"
# Fallback: direct download page that redirects
DOWNLOAD_URL = "https://www.cursor.com/api/download?platform=linux-x64&releaseTrack=stable"


def get_latest_version() -> dict:
    """Query Cursor's update API for the latest version info."""
    # Try the update API first
    try:
        req = urllib.request.Request(
            UPDATE_URL,
            headers={"User-Agent": "cursor-ide-bin-updater/1.0"},
        )
        with urllib.request.urlopen(req, timeout=30) as resp:
            data = json.loads(resp.read().decode())
            if "version" in data and "name" in data:
                return {
                    "version": data["name"],
                    "commit": data["version"],
                    "url": data.get("url", ""),
                }
    except Exception as e:
        print(f"Update API failed ({e}), trying fallback...", file=sys.stderr)

    # Fallback: follow the download redirect to extract version from URL
    try:
        req = urllib.request.Request(
            DOWNLOAD_URL,
            headers={"User-Agent": "cursor-ide-bin-updater/1.0"},
        )
        with urllib.request.urlopen(req, timeout=30) as resp:
            final_url = resp.url
            # URL pattern: .../production/<commit>/linux/x64/deb/amd64/deb/cursor_<ver>_amd64.deb
            m = re.search(
                r"/production/([a-f0-9]+)/linux/x64/deb/amd64/deb/cursor_([0-9.]+)_amd64\.deb",
                final_url,
            )
            if m:
                return {
                    "version": m.group(2),
                    "commit": m.group(1),
                    "url": final_url,
                }
    except Exception as e:
        print(f"Fallback also failed: {e}", file=sys.stderr)

    # Last resort: try the JSON API endpoint that the original updater uses
    try:
        api_url = "https://downloads.cursor.com/production/latest/linux/x64/deb/amd64/deb"
        req = urllib.request.Request(
            api_url,
            headers={"User-Agent": "cursor-ide-bin-updater/1.0"},
            method="HEAD",
        )
        # Don't follow redirects manually
        opener = urllib.request.build_opener(NoRedirectHandler())
        resp = opener.open(req, timeout=30)
        location = resp.headers.get("Location", "")
        m = re.search(
            r"/production/([a-f0-9]+)/linux/x64/deb/amd64/deb/cursor_([0-9.]+)_amd64\.deb",
            location,
        )
        if m:
            return {
                "version": m.group(2),
                "commit": m.group(1),
                "url": location,
            }
    except Exception:
        pass

    print("ERROR: Could not determine latest Cursor version from any endpoint.", file=sys.stderr)
    sys.exit(1)


class NoRedirectHandler(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None

    def http_error_302(self, req, fp, code, msg, headers):
        return fp

    http_error_301 = http_error_303 = http_error_307 = http_error_302


def get_current_version(pkgbuild: Path) -> tuple[str, str]:
    """Extract current pkgver and _commit from PKGBUILD."""
    text = pkgbuild.read_text()
    ver_m = re.search(r"^pkgver=(.+)$", text, re.MULTILINE)
    commit_m = re.search(r"^_commit=(.+)$", text, re.MULTILINE)
    if not ver_m or not commit_m:
        print("ERROR: Could not parse pkgver/_commit from PKGBUILD", file=sys.stderr)
        sys.exit(1)
    return ver_m.group(1).strip(), commit_m.group(1).strip()


def compute_sha256(url: str) -> str:
    """Download file and compute SHA256."""
    print(f"Downloading {url} for checksum...", file=sys.stderr)
    req = urllib.request.Request(url, headers={"User-Agent": "cursor-ide-bin-updater/1.0"})
    sha = hashlib.sha256()
    with urllib.request.urlopen(req, timeout=300) as resp:
        while True:
            chunk = resp.read(65536)
            if not chunk:
                break
            sha.update(chunk)
    return sha.hexdigest()


def update_pkgbuild(pkgbuild: Path, new_ver: str, new_commit: str, new_sha256: str):
    """Update PKGBUILD with new version, commit, and checksum."""
    text = pkgbuild.read_text()

    text = re.sub(r"^pkgver=.+$", f"pkgver={new_ver}", text, flags=re.MULTILINE)
    text = re.sub(r"^pkgrel=.+$", "pkgrel=1", text, flags=re.MULTILINE)
    text = re.sub(r"^_commit=.+$", f"_commit={new_commit}", text, flags=re.MULTILINE)

    # Update sha256sums — first entry is the deb
    text = re.sub(
        r"(sha256sums=\(')[^']*(')",
        rf"\g<1>{new_sha256}\2",
        text,
    )

    pkgbuild.write_text(text)


def generate_srcinfo(pkgbuild_dir: Path) -> str:
    """Generate .SRCINFO from PKGBUILD using makepkg."""
    result = subprocess.run(
        ["makepkg", "--printsrcinfo"],
        cwd=pkgbuild_dir,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        print(f"makepkg --printsrcinfo failed: {result.stderr}", file=sys.stderr)
        sys.exit(1)
    return result.stdout


def main():
    parser = argparse.ArgumentParser(description="Update cursor-ide-bin PKGBUILD")
    parser.add_argument(
        "--check", action="store_true",
        help="Only check, don't update. Prints JSON with latest version info.",
    )
    parser.add_argument(
        "--update", action="store_true",
        help="Update PKGBUILD if a new version is available.",
    )
    parser.add_argument(
        "--pkgbuild", type=Path, default=Path(__file__).parent / "PKGBUILD",
        help="Path to PKGBUILD (default: ./PKGBUILD)",
    )
    parser.add_argument(
        "--skip-checksum", action="store_true",
        help="Skip SHA256 download (use SKIP). Faster but less secure.",
    )
    args = parser.parse_args()

    if not args.check and not args.update:
        parser.print_help()
        sys.exit(1)

    latest = get_latest_version()
    current_ver, current_commit = get_current_version(args.pkgbuild)

    print(f"Current: {current_ver} ({current_commit[:12]}...)", file=sys.stderr)
    print(f"Latest:  {latest['version']} ({latest['commit'][:12]}...)", file=sys.stderr)

    if args.check:
        is_new = latest["version"] != current_ver or latest["commit"] != current_commit
        out = {
            "current_version": current_ver,
            "latest_version": latest["version"],
            "latest_commit": latest["commit"],
            "update_available": is_new,
        }
        print(json.dumps(out, indent=2))
        return

    if args.update:
        if latest["version"] == current_ver and latest["commit"] == current_commit:
            print("Already up-to-date.", file=sys.stderr)
            sys.exit(2)

        # Compute checksum
        deb_url = (
            f"https://downloads.cursor.com/production/{latest['commit']}"
            f"/linux/x64/deb/amd64/deb/cursor_{latest['version']}_amd64.deb"
        )
        if args.skip_checksum:
            sha256 = "SKIP"
        else:
            sha256 = compute_sha256(deb_url)
            print(f"SHA256: {sha256}", file=sys.stderr)

        update_pkgbuild(args.pkgbuild, latest["version"], latest["commit"], sha256)
        print(f"PKGBUILD updated to {latest['version']}", file=sys.stderr)

        # Regenerate .SRCINFO
        srcinfo_path = args.pkgbuild.parent / ".SRCINFO"
        try:
            srcinfo = generate_srcinfo(args.pkgbuild.parent)
            srcinfo_path.write_text(srcinfo)
            print(".SRCINFO regenerated", file=sys.stderr)
        except Exception as e:
            print(f"Warning: Could not regenerate .SRCINFO: {e}", file=sys.stderr)
            print("You may need to run 'makepkg --printsrcinfo > .SRCINFO' manually.", file=sys.stderr)


if __name__ == "__main__":
    main()

File: update.ts
import { createHash } from "node:crypto";
import { readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const UPDATE_URL =
	"https://cursor.com/api/download?platform=linux-x64&releaseTrack=stable";

const __dirname = dirname(fileURLToPath(import.meta.url));

interface VersionInfo {
	version: string;
	commit: string;
	url: string;
}

async function getLatestVersion(): Promise<VersionInfo> {
	try {
		const resp = await fetch(UPDATE_URL, {
			headers: { "User-Agent": "cursor-ide-updater/1.0" },
			signal: AbortSignal.timeout(30_000),
		});
		const data = await resp.json();

		if (data.version && data.commitSha) {
			return {
				version: data.version,
				commit: data.commitSha,
				url: data.debUrl ?? "",
			};
		}

		if (data.name && data.version) {
			return {
				version: data.name,
				commit: data.version,
				url: data.url ?? "",
			};
		}
	} catch (e) {
		console.error(`Update API failed: ${e}`);
	}

	console.error("ERROR: Could not determine latest Cursor version.");
	return process.exit(1);
}

function getCurrentVersion(pkgbuildPath: string): {
	version: string;
	commit: string;
} {
	const text = readFileSync(pkgbuildPath, "utf-8");
	const verMatch = text.match(/^pkgver=(.+)$/m);
	const commitMatch = text.match(/^_commit=(.+)$/m);
	if (!verMatch || !commitMatch) {
		console.error("ERROR: Could not parse pkgver/_commit from PKGBUILD");
		process.exit(1);
	}
	return { version: verMatch[1].trim(), commit: commitMatch[1].trim() };
}

async function computeSha256(url: string): Promise<string> {
	console.error(`Downloading ${url} for checksum...`);
	const resp = await fetch(url, {
		headers: { "User-Agent": "cursor-ide-updater/1.0" },
		signal: AbortSignal.timeout(300_000),
	});
	if (!resp.ok || !resp.body) {
		throw new Error(`Download failed: ${resp.status} ${resp.statusText}`);
	}
	const hash = createHash("sha256");
	for await (const chunk of resp.body) {
		hash.update(chunk);
	}
	return hash.digest("hex");
}

function updatePkgbuild(
	pkgbuildPath: string,
	newVer: string,
	newCommit: string,
	newSha256: string,
): void {
	let text = readFileSync(pkgbuildPath, "utf-8");
	text = text.replace(/^pkgver=.+$/m, `pkgver=${newVer}`);
	text = text.replace(/^pkgrel=.+$/m, "pkgrel=1");
	text = text.replace(/^_commit=.+$/m, `_commit=${newCommit}`);
	text = text.replace(/(sha256sums=\(')[^']*(')/, `$1${newSha256}$2`);
	writeFileSync(pkgbuildPath, text);
}

function generateSrcinfo(pkgbuildPath: string): string {
	const text = readFileSync(pkgbuildPath, "utf-8");

	const field = (re: RegExp): string => {
		const m = text.match(re);
		return m ? m[1].trim() : "";
	};

	const pkgver = field(/^pkgver=(.+)$/m);
	const commit = field(/^_commit=(.+)$/m);

	// Expand bash variables in a string
	const expand = (s: string): string =>
		s
			.replace(/\$\{pkgver\}|\$pkgver/g, pkgver)
			.replace(/\$\{_commit\}|\$_commit/g, commit);

	const arrayField = (name: string): string[] => {
		const re = new RegExp(`^${name}=\\(([\\s\\S]*?)\\)`, "m");
		const m = text.match(re);
		if (!m) return [];

		const items: string[] = [];
		for (const line of m[1].split("\n")) {
			const stripped = line.replace(/#.*/, "").trim();
			if (!stripped) continue;
			// Quoted strings are kept whole (may contain spaces)
			if (/^['"]/.test(stripped)) {
				items.push(expand(stripped.replace(/^['"]|['"]$/g, "")));
			} else {
				// Unquoted: split on whitespace (e.g. "!strip !debug")
				for (const tok of stripped.split(/\s+/)) {
					if (tok) items.push(expand(tok));
				}
			}
		}
		return items;
	};

	const pkgname = field(/^pkgname=(.+)$/m);
	const pkgrel = field(/^pkgrel=(.+)$/m);
	const pkgdesc = field(/^pkgdesc=['"](.*?)['"]$/m);
	const url = field(/^url=['"](.*?)['"]$/m);
	const arch = arrayField("arch");
	const license = arrayField("license");
	const makedepends = arrayField("makedepends");
	const depends = arrayField("depends");
	const optdepends = arrayField("optdepends");
	const provides = arrayField("provides");
	const options = arrayField("options");
	const noextract = arrayField("noextract");
	const sources = arrayField("source");
	const sha256sums = arrayField("sha256sums");

	const lines: string[] = [];
	const global = (key: string, val: string) =>
		lines.push(`\t${key} = ${val}`);

	lines.push(`pkgbase = ${pkgname}`);
	global("pkgdesc", pkgdesc);
	global("pkgver", pkgver);
	global("pkgrel", pkgrel);
	global("url", url);
	for (const a of arch) global("arch", a);
	for (const l of license) global("license", l);
	for (const m of makedepends) global("makedepends", m);
	for (const d of depends) global("depends", d);
	for (const o of optdepends) global("optdepends", o);
	for (const p of provides) global("provides", p);
	for (const n of noextract) global("noextract", n);
	for (const o of options) global("options", o);
	for (const s of sources) global("source", s);
	for (const s of sha256sums) global("sha256sums", s);
	lines.push("");
	lines.push(`pkgname = ${pkgname}`);

	return lines.join("\n") + "\n";
}

async function main(): Promise<void> {
	const args = process.argv.slice(2);
	const check = args.includes("--check");
	const update = args.includes("--update");
	const skipChecksum = args.includes("--skip-checksum");

	if (!check && !update) {
		console.error(
			"Usage: npx tsx update.ts [--check | --update] [--skip-checksum]",
		);
		process.exit(1);
	}

	const pkgbuildArg = args.find((a) => a.startsWith("--pkgbuild="));
	const pkgbuildPath = pkgbuildArg
		? pkgbuildArg.split("=")[1]
		: join(__dirname, "PKGBUILD");

	const latest = await getLatestVersion();
	const current = getCurrentVersion(pkgbuildPath);

	console.error(
		`Current: ${current.version} (${current.commit.slice(0, 12)}...)`,
	);
	console.error(
		`Latest:  ${latest.version} (${latest.commit.slice(0, 12)}...)`,
	);

	if (check) {
		const isNew =
			latest.version !== current.version || latest.commit !== current.commit;
		const out = {
			current_version: current.version,
			latest_version: latest.version,
			latest_commit: latest.commit,
			update_available: isNew,
		};
		console.log(JSON.stringify(out, null, 2));
		return;
	}

	if (update) {
		if (
			latest.version === current.version &&
			latest.commit === current.commit
		) {
			console.error("Already up-to-date.");
			process.exit(2);
		}

		const debUrl = `https://downloads.cursor.com/production/${latest.commit}/linux/x64/deb/amd64/deb/cursor_${latest.version}_amd64.deb`;
		let sha256: string;
		if (skipChecksum) {
			sha256 = "SKIP";
		} else {
			sha256 = await computeSha256(debUrl);
			console.error(`SHA256: ${sha256}`);
		}

		updatePkgbuild(pkgbuildPath, latest.version, latest.commit, sha256);
		console.error(`PKGBUILD updated to ${latest.version}`);

		const srcinfoPath = join(dirname(pkgbuildPath), ".SRCINFO");
		try {
			const srcinfo = generateSrcinfo(pkgbuildPath);
			writeFileSync(srcinfoPath, srcinfo);
			console.error(".SRCINFO regenerated");
		} catch (e) {
			console.error(`Warning: Could not regenerate .SRCINFO: ${e}`);
			console.error(
				"You may need to run 'makepkg --printsrcinfo > .SRCINFO' manually.",
			);
		}
	}
}

main();
